src/puppetlabs/trapperkeeper/authorization/rules.clj

```diff
@@ -125,67 +121,90 @@
+    (str "Forbidden request: " from " access to " path " (method " method ")"
          (if-let [file (:file rule)] (str " at " file ":" (:line rule)))
          (format " (authentic: %s) " authentic?)
          (format "denied by rule '%s'." (:name rule)))))
 
-(schema/defn sort-rules :- Rules
-  "Sorts the rules based on their :sort-order, and then their :name if they
-   have the same sort order value."
-  [rules :- Rules]
-  (sort-by (juxt :sort-order :name) rules))
+(schema/defn allow-request :- AuthorizationResult
+  "Logs debugging information about the request and rule at the TRACE level
+   and returns an authorized authorization result with the provided message."
+  [request :- ring/Request
+   rule :- Rule
+   message :- schema/Str]
+  (log/tracef
+   "Request to '%s' from '%s' handled by rule '%s' - request allowed"
+   (:uri request) (requestor request) (:name rule))
+  {:authorized true
+   :message message})
 
 (schema/defn deny-request :- AuthorizationResult
-  "Logs the reason message at ERROR level and
-   returns an unauthorized authorization result."
-  [reason :- schema/Str]
+  "Logs debugging information about the request and rule at the TRACE level
+   as well as the reason for denial at the ERROR level, and returns an
+   unauthorized authorization result with the provided reason message."
+  [request :- ring/Request
+   rule :- (schema/maybe Rule)
+   reason :- schema/Str]
+  (if rule
+    (log/tracef
+     "Request to '%s' from '%s' handled by rule '%s' - request denied"
+     (:uri request) (requestor request) (:name rule))
+    (log/tracef
+     "Request to '%s' from '%s' did not match any rules - request denied"
+     (:uri request) (requestor request)))
   (log/error reason)
   {:authorized false
    :message reason})
 
-;; Rules creation
-
-(def empty-rules [])
-
-(schema/defn add-rule
-  [rules :- Rules
-   rule :- Rule]
-  (conj rules rule))
-
 ;; Rules check
 
 (schema/defn allowed? :- AuthorizationResult
   "Checks if a request is allowed access given the list of rules. Rules
    will be checked in the given order; use `sort-rules` to first sort them."
-  [rules :- Rules
-   request :- ring/Request
-   name :- schema/Str]
+  [rules :- [Rule]
+   request :- ring/Request]
   (if-let [{:keys [rule matches]} (some #(match? % request) rules)]
     (if (true? (:allow-unauthenticated rule))
-      {:authorized true :message "allow-unauthenticated is true - allowed"}
-      (if (and (true? (ring/authorized-authentic? request)) ; authenticated?
-            (acl/allowed? (:acl rule) name (:remote-addr request) matches))
-        {:authorized true :message ""}
-        (deny-request (request->description request name rule))))
-    (deny-request "global deny all - no rules matched")))
+      (allow-request request rule "allow-unauthenticated is true - allowed")
+      (if (and (true? (ring/authorized-authentic? request))
+               (acl/allowed? (:acl rule) (or (ring/authorized-name request) "")
```

Comment:
If we leave it as an empty string and not potentially nil, nothing has to change here, right?  Ok, I'll shut up now...